src/MongoDB.Driver/Linq/Translators/FindProjectionTranslator.cs

```diff
@@ -192,7 +218,12 @@ private static IReadOnlyList<IFieldExpression> GetUniqueFieldsByHierarchy(IEnume
                 var referenceGroup = referenceGroups.Dequeue();
                 if (!skippedFields.Contains(referenceGroup.Key))
                 {
-                    var hierarchicalReferenceGroups = referenceGroups.Where(x => x.Key.StartsWith(referenceGroup.Key));
+                    var hierarchicalReferenceGroups = referenceGroups
+                        .Where(x =>
+                        {
+                            return x.Key.Count(c => c == '.') != referenceGroup.Key.Count(c => c == '.') && // otherwise, fields are on the same nesting level
```

Comment:
if field `A` and field `B` has the same `dot` numbers, it means that they are on the same nesting level. Nothing should be skipped.

NOTE: if we have projection like `p => new { p.C.E, p.C.E }`, the length of `referenceGroups` in the line `215` will be `1`, because we make grouping by `FieldName`. So, the projection document will be also correct.